waspc/src/Wasp/Generator/Setup.hs

```diff
@@ -1,22 +1,82 @@
 module Wasp.Generator.Setup
-  ( setup,
+  ( runSetup,
   )
 where
 
 import Control.Concurrent (Chan, newChan, readChan)
 import Control.Concurrent.Async (concurrently)
-import StrongPath (Abs, Dir, Path')
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString.Lazy as B
+import StrongPath (Abs, Dir, File', Path', Rel, relfile, (</>))
+import qualified StrongPath as SP
+import System.Directory (doesFileExist)
 import System.Exit (ExitCode (..))
+import Wasp.AppSpec (AppSpec)
 import Wasp.Generator.Common (ProjectRootDir)
+import qualified Wasp.Generator.DbGenerator as DbGenerator
 import qualified Wasp.Generator.Job as J
 import Wasp.Generator.Job.IO (printPrefixedJobMessage)
-import Wasp.Generator.ServerGenerator.Setup (setupServer)
-import Wasp.Generator.WebAppGenerator.Setup (setupWebApp)
+import Wasp.Generator.Monad (GeneratorError (..), GeneratorWarning (..))
+import qualified Wasp.Generator.PackageJsonGenerator as PJG
+import Wasp.Generator.ServerGenerator as SG
+import qualified Wasp.Generator.ServerGenerator.Setup as SSetup
+import Wasp.Generator.WebAppGenerator as WG
+import qualified Wasp.Generator.WebAppGenerator.Setup as WSetup
 
-setup :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
-setup projectDir = do
+runSetup :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
+runSetup spec dstDir = do
+  (npmInstallWarnings, npmInstallErrors) <- runNpmInstall spec dstDir
+  (dbGeneratorWarnings, dbGeneratorErrors) <- DbGenerator.postWriteDbGeneratorActions spec dstDir
+  return (npmInstallWarnings ++ dbGeneratorWarnings, npmInstallErrors ++ dbGeneratorErrors)
+
+-- | Run npm install if needed
+runNpmInstall :: AppSpec -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
+runNpmInstall spec dstDir = do
+  let appSpecAllPackageJsonDependencies = PJG.getAllPackageJsonDependencies spec SG.waspPackageJsonDependencies WG.waspPackageJsonDependencies
+  case appSpecAllPackageJsonDependencies of
+    Nothing -> return ([], [GenericGeneratorError "npm install failed due to conflict between user and wasp dependencies."])
+    Just allDependencies -> executeNpmInstall allDependencies dstDir
+
+-- Run npm install for desired AppSpec dependencies if needed
+executeNpmInstall :: PJG.AllPackageJsonDependencies -> Path' Abs (Dir ProjectRootDir) -> IO ([GeneratorWarning], [GeneratorError])
+executeNpmInstall appSpecAllPackageJsonDependencies dstDir = do
+  needed <- isNpmInstallNeeded appSpecAllPackageJsonDependencies dstDir
+  npmInstallResult <-
+    if needed
+      then installNpmDependencies dstDir
```

Comment:
If `npm install` fails, the npmDependencies won't be updated. Therefore the system will continue to assume you need to rerun `npm install` to get into the right state, which appears to be the desired behavior to me.

Reply:
I have added a removeFile in here in anyway, because it's conceivable that `npm install` breaks in such a way that it left `node_modules` in weird state. I am not sure that this would resolve it but I don't think it will hurt.